src/python/pants/backend/python/target_types_rules.py

```diff
@@ -135,29 +142,124 @@ async def inject_pex_binary_entry_point_dependency(
 # -----------------------------------------------------------------------------------------------
 
 
+@rule(desc="Determining the entry points for a `python_distribution` target")
+async def resolve_python_distribution_entry_points(
+    request: ResolvePythonDistributionEntryPointsRequest,
+) -> ResolvedPythonDistributionEntryPoints:
+    field_value = request.entry_points_field.value
+    if field_value is None:
+        return ResolvedPythonDistributionEntryPoints(None)
+
+    address = request.entry_points_field.address
+    entry_points: dict[str, dict[str, EntryPoint]] = defaultdict(dict)
+
+    for section, values in field_value.items():
+        for name, ref in values.items():
+            if ref.startswith(":") or "/" in ref:
+                targets = await Get(Targets, UnparsedAddressInputs([ref], owning_address=address))
+                if not targets:
+                    raise InvalidFieldException(
+                        f'The "entry_points" target address {repr(ref)} does not resolve to any known target.'
+                    )
+
+                for target in targets:
+                    if target.has_field(PexEntryPointField):
+                        binary_entry_point = await Get(
+                            ResolvedPexEntryPoint,
+                            ResolvePexEntryPointRequest(target[PexEntryPointField]),
+                        )
+                        ep = binary_entry_point.val
+                        break
+                else:
+                    raise InvalidFieldException(
+                        f'The "entry_points" target address {repr(ref)} does not resolve to a pex_binary() target, but:'
+                        + ", ".join(str(t) for t in targets)
+                    )
```

Comment:
As discussed over DM, this has been confusing me a lot on whether you need to use a `pex_binary` or not. 

I think it makes sense to always require `pex_binary`. We can't reliably convert any other target type like `python_library` into a module. We can only do that with `pex_binary` because it has an `entry_point` field.

--

There's additional nuance: the `console_scripts` category `console_scripts` requires the `:func` part to be defined. So, for `.with_binaries()`, we've been requiring that you specify the `:func` part in the original `pex_binary`. I think we want to eagerly validate that here too when the section is `console_scripts`:

https://github.com/pantsbuild/pants/blob/ddc3c3d791178a610a9b8988cb74ac072bdac9c0/src/python/pants/backend/python/goals/setup_py.py#L584-L590